### PR TITLE
[#188] Added an OperationId decorator that overrides inferred id

### PR DESCRIPTION
**Pull request by redlock**

### CODE REVIEW

File: src/metadataGeneration/methodGenerator.ts [hunk 1]
```diff
@@ -200,6 +201,21 @@ export class MethodGenerator {
...
+      return undefined;
+    }
+    if (opDecorators.length > 1) {
+      throw new GenerateMetadataError(`Only one OperationId decorator allowed in '${this.getCurrentLocation}' method.`);
+    }
+
+    const decorator = opDecorators[0];
+    const expression = decorator.parent as ts.CallExpression;
+    const ops = expression.arguments.map((a: any) => a.text as string);
+    return ops[0]
```
Comment (endor) [inline src/metadataGeneration/methodGenerator.ts]: `;` missing

File: src/swagger/specGenerator.ts [hunk 2]
```diff
@@ -100,7 +100,11 @@ export class SpecGenerator {
     const pathMethod: Swagger.Operation = pathObject[method.method] = this.buildOperation(controllerName, method);
     pathMethod.description = method.description;
     pathMethod.summary = method.summary;
-    pathMethod.tags = method.tags;
+    
+    pathMethod.tags = [controllerName].concat(method.tags || [])
```
Comment (endor) [inline src/swagger/specGenerator.ts]: What does this do?

File: src/swagger/specGenerator.ts [hunk 3]
```diff
@@ -100,7 +100,11 @@ export class SpecGenerator {
     const pathMethod: Swagger.Operation = pathObject[method.method] = this.buildOperation(controllerName, method);
     pathMethod.description = method.description;
     pathMethod.summary = method.summary;
-    pathMethod.tags = method.tags;
+    
+    pathMethod.tags = [controllerName].concat(method.tags || []) 
+    
+    // Use operationId tag otherwise fallback to generated. Warning: This doesn't check uniqueness.
```
Comment (endor) [inline src/swagger/specGenerator.ts]: Should uniqueness be checked?

**Comment by endor** [comment]
It would be good to:

* Add unit tests to show how this works (one example where it's overridden and one where it's not)
* Add documentation to the README in the decorators section
* Squash the commits

File: src/swagger/specGenerator.ts (hunk 2 above)
Comment (redlock) [reply src/swagger/specGenerator.ts]: This creates a default tag based on the controller name. I'm using it to group methods by controller name. I will remove it from the pull request as it is another modification that isn't related to operationId.

File: src/swagger/specGenerator.ts (hunk 3 above)
Comment (endor) [reply src/swagger/specGenerator.ts]: Okay, let's go without it for now. Then if people later want it, we can still add it in.

File: src/swagger/specGenerator.ts [hunk 4]
```diff
@@ -100,7 +100,11 @@ export class SpecGenerator {
     const pathMethod: Swagger.Operation = pathObject[method.method] = this.buildOperation(controllerName, method);
     pathMethod.description = method.description;
     pathMethod.summary = method.summary;
-    pathMethod.tags = method.tags;
+    
+    pathMethod.tags = method.tags // [controllerName].concat(method.tags || [])
```
Comment (endor) [inline src/swagger/specGenerator.ts]: `;` missing. And I think we can remove the comment.

**Comment by redlock** [comment]
So I've handled all the points except the unit test. I'm just not sure where to place it, and I'm not familiar with the test suite/configuration. I'd appreciate if someone more familiar with the tests can add it. Thanks.

**Pull request closed without merging**